Support newer java versions

davecromberge opened the pull request:
This is a major rework of the memory project structure that provides support for Java 8, as well as Java 9 through 13.  

The changes leverage improvements in the Java ecosystem such as the Java Module System (JPMS), and multi-release JARs (MR-Jars).  

The PR here is a consolidation of work conducted together with @leerho over several months, and documentation has been provided to the rationale for the changes.

For further information, see this issue: https://github.com/apache/datasketches-memory/issues/121

davecromberge committed:
Set version to 2.0.0-SNAPSHOT

davecromberge committed:
Introduce maven multi-module configuration

The purpose of this change is to create a release that is usable
from multilpe Java versions.  This initially supports parity
with the current release, which is now baselined to Java8.

davecromberge committed:
Split java8 maven module

Separate the java8 source and tests into separate maven
modules.

davecromberge committed:
Add datasketches-memory module

This module depends on all other modules in the project, and is used
to assemble JARs for installation to a local repository or deployment
to a remote repository.

davecromberge committed:
Add java9 module scaffolding

davecromberge committed:
Add support for Multi-Release JAR assembly

Extend the assembled JAR to allow multiple, Java-release-specific
versions of class files to coexist in a single archive.

See JEP-238.

davecromberge committed:
Add support for Java 9 and Java 10

Extract the platform specific features to the java9 maven module:
- reference cleaner
- nio bits fields
- virtual machine memory fields

This makes use of the new MR-jar to coexist with the java8 equivalents.

davecromberge committed:
Add support for Java Platform Module System

Fixes compiler error where internals are not accessible to java9
related code.  Furthermore, a JPMS module is defined for
the datasketches-memory, that can be used to narrow any additional
compiler or runtime arguments for the JPMS.

davecromberge committed:
Add datasketches-memory-tests maven module

This module runs the test suite against the packaged JAR, using the
java version defined by the environment that runs maven.

davecromberge committed:
Use dynamic profiles when testing the assembled JAR

This prevents unsupported runtime flags being used in Java8
during test execution using the Maven surefire plugin.

davecromberge committed:
Add support for Java 11

davecromberge committed:
Update IDE documentation

davecromberge committed:
Remove slf4j dependency

davecromberge committed:
Use aggregate test coverage report for coveralls

davecromberge committed:
Use Github actions matrix build

This allows us to verify that the MR-Jar works across various
JVMs by running the unit test suite across each Java version.

davecromberge committed:
Add java 9 toolchain

davecromberge committed:
Adjust activation string for strict profile

davecromberge committed:
Add initial LGTM project settings

leerho commented on README.md:
```diff
@@ -36,13 +36,49 @@ Please visit the main [DataSketches website](https://datasketches.apache.org) fo
-### JDK8/Hotspot is required to compile
-This DataSketches component is pure Java and you must compile using JDK 8 with Hotspot.
+### JDK versions required to compile
+This DataSketches component is pure Java and requires the following JDKs to compile:
```
I think you mean "one of the following".  (certainly not all of them simultaneously :) )

davecromberge replied:
I am not sure how to clarify this documentation - as the Maven toolchain build requires these three java versions to be present on the developer's machine in order to build the project successfully.  The end-user does not require all java versions to be present simultaneously.

Do you think that this needs further explanation in the README?

davecromberge committed:
Correct toolchains location

davecromberge committed:
Checkstyle fixes and formatting changes

davecromberge committed:
Update dependencies

davecromberge committed:
Remove unused add-opens from assembly tests

davecromberge committed:
Introduce transient dependency scope

davecromberge committed:
Support maven nexus deployment

davecromberge committed:
Use maven aggregation over inheritance

This prevents the memory assembly artifacts depending on the root
module.  However, this configuration duplicates some properties
between the root and assembly POM files.

davecromberge committed:
Additional matrix test versions

davecromberge committed:
Remove additonal matrix tests

davecromberge committed:
Explicit toolchain version in github actions

davecromberge committed:
Instructions for javadoc command

davecromberge committed:
Remove maven help plugin and align poms
